### PR TITLE
 Implement discovery registry through Mysterium Communication

### PR DESCRIPTION
**Pull request by Waldz**
Expected behaviour:
- `--discovery.type` works in such a way, that Provider register new proposals to NAT broker to subject `*.proposal-register`
- All consumer who are interested in new proposals can listen to this queue and store them to local storage
- Also there is no difference between proposal registration and pinging, Provider pings his alive proposal actively until it's alive

Changes made:
- Discovery related code moved to separate package `core/discovery`
- Introduced `discovery.Registry` for abstraction of proposals announcing 
- 2 different Discovery providers introduced: `api` and `broker` 
- Discovery provider can be chosen by `myst` arguments `--discovery.type` and `--discovery.address`

Screen Shot 2019-05-10 at 12 52 50

### CODE REVIEW

**Comment by vkuznecovas** [comment]
Could you please explain why is this even needed?

**Comment by Waldz** [comment]
I am implementing:
- Golang discovery

File: cmd/di.go [hunk 1]
```diff
@@ -438,8 +438,8 @@ func (di *Dependencies) bootstrapNetworkComponents(options node.OptionsNetwork)
 	}
 
 	//override defined values one by one from options
-	if options.DiscoveryAPIAddress != metadata.DefaultNetwork.DiscoveryAPIAddress {
-		network.DiscoveryAPIAddress = options.DiscoveryAPIAddress
+	if options.APIAddress != metadata.DefaultNetwork.APIAddress {
```
Comment (tadaskay) [inline cmd/di.go]: What kind of API is it? Further, what kind of address is it? If you name it `MysteriumApiUrl`, then all ambiguities are gone

**Comment by tadaskay** [comment]
If I understand correctly, you should be able to discover proposals from multiple nodes if we're really to decentralize this aspect. Not only point discovery to a different URL. Is that correct?

File: cmd/di.go (hunk 1 above)
Comment (Waldz) [reply cmd/di.go]: Comment addressed

File: core/discovery/communication/registry.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		communication.NewCodecJSON(),
+		"*",
+	)
+}
+
+type registry struct {
+	sender communication.Sender
+}
+
+// NewRegistry create instance if Broker registry
```
Comment (soffokl) [inline core/discovery/communication/registry.go]:
```suggestion
// NewRegistry create an instance of Broker registry
```

File: core/discovery/discovery.go [hunk 3]
```diff
@@ -44,6 +45,45 @@ const (
...
 const logPrefix = "[discovery] "
 
+// Discovery structure holds discovery service state
+type Discovery struct {
+	identityRegistry            identity_registry.IdentityRegistry
+	ownIdentity                 identity.Identity
+	identityRegistration        identity_registry.RegistrationDataProvider
+	proposalRegistry            ProposalRegistry
+	signerCreate                identity.SignerFactory
+	signer                      identity.Signer
```
Comment (soffokl) [inline core/discovery/discovery.go]: Do we really need both `identity.Signer` and `identity.SignerFactory`?

File: core/discovery/discovery.go [hunk 4]
```diff
@@ -44,6 +45,45 @@ const (
...
+	proposal                    market.ServiceProposal
+	statusChan                  chan Status
+	status                      Status
+	proposalAnnouncementStopped *sync.WaitGroup
+	unsubscribe                 func()
+	stop                        func()
+
+	sync.RWMutex
```
Comment (soffokl) [inline core/discovery/discovery.go]: Embedding mutex will allow accessing `Lock()` and `Unlock()` outside.

File: core/discovery/discovery.go [hunk 5]
```diff
@@ -44,6 +45,45 @@ const (
...
+	identityRegistration        identity_registry.RegistrationDataProvider
+	proposalRegistry            ProposalRegistry
+	signerCreate                identity.SignerFactory
+	signer                      identity.Signer
+	proposal                    market.ServiceProposal
+	statusChan                  chan Status
+	status                      Status
+	proposalAnnouncementStopped *sync.WaitGroup
+	unsubscribe                 func()
+	stop                        func()
```
Comment (soffokl) [inline core/discovery/discovery.go]: It looks like `stop` and `unsubscribe` are doing the same things.

File: core/discovery/discovery.go [hunk 6]
```diff
@@ -44,6 +45,45 @@ const (
...
+		identityRegistry:            identityRegistry,
+		identityRegistration:        identityRegistration,
+		proposalRegistry:            proposalRegistry,
+		signerCreate:                signerCreate,
+		statusChan:                  make(chan Status),
+		status:                      StatusUndefined,
+		proposalAnnouncementStopped: &sync.WaitGroup{},
+		unsubscribe:                 func() {},
+		stop:                        func() {},
+		RWMutex:                     sync.RWMutex{},
```
Comment (soffokl) [inline core/discovery/discovery.go]: We can skip this line since it's a default value.

File: core/discovery/discovery.go [hunk 7]
```diff
@@ -44,6 +45,45 @@ const (
...
+) *Discovery {
+	return &Discovery{
+		identityRegistry:            identityRegistry,
+		identityRegistration:        identityRegistration,
+		proposalRegistry:            proposalRegistry,
+		signerCreate:                signerCreate,
+		statusChan:                  make(chan Status),
+		status:                      StatusUndefined,
+		proposalAnnouncementStopped: &sync.WaitGroup{},
+		unsubscribe:                 func() {},
```
Comment (soffokl) [inline core/discovery/discovery.go]: Why we need this useless functions `unsubscribe` and `stop`?
If it's not used now, we should not add it.

File: core/discovery/interface.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2018 The "MysteriumNetwork/node" Authors.
```
Comment (soffokl) [inline core/discovery/interface.go]:
```suggestion
 * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
```

File: cmd/di.go [hunk 9]
```diff
@@ -146,6 +150,8 @@ type Dependencies struct {
 	IdentityRegistry     identity_registry.IdentityRegistry
 	IdentityRegistration identity_registry.RegistrationDataProvider
 
+	DiscoveryFactory service.DiscoveryFactory
```
[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/di.go]: Existing code, not my scope

File: core/discovery/communication/registry.go (hunk 2 above)
Comment (Waldz) [reply core/discovery/communication/registry.go]: Existing code, not my scope

File: core/discovery/discovery.go (hunk 3 above)
Comment (Waldz) [reply core/discovery/discovery.go]: Existing code, not my scope

File: core/discovery/discovery.go (hunk 5 above)
Comment (Waldz) [reply core/discovery/discovery.go]: Existing code, not my scope

File: core/discovery/discovery.go (hunk 7 above)
Comment (Waldz) [reply core/discovery/discovery.go]: Existing code, not my scope.
And yes, it's being used

File: cmd/di.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply cmd/di.go]: But.. you just introduced this?

File: core/discovery/communication/registry.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+// NewRegistry create an instance of Broker registry
+func NewRegistry(sender communication.Sender) *registry {
+	return &registry{
+		sender: sender,
+	}
+}
+
+// RegisterProposal registers service proposal to discovery service
+func (registry *registry) RegisterProposal(proposal market.ServiceProposal, signer identity.Signer) error {
```
Comment (vkuznecovas) [inline core/discovery/communication/registry.go]:
```suggestion
func (registry *registry) RegisterProposal(proposal market.ServiceProposal, _ identity.Signer) error {
```

File: core/discovery/communication/registry.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	message := &registerMessage{Proposal: proposal}
+	return registry.sender.Send(&registerProducer{message: message})
+}
+
+// UnregisterProposal unregisters a service proposal when client disconnects
+func (registry *registry) UnregisterProposal(proposal market.ServiceProposal, signer identity.Signer) error {
```
Comment (vkuznecovas) [inline core/discovery/communication/registry.go]: Do explicitly omit the signer if you're not going to use it.

File: core/discovery/communication/registry.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	message := &unregisterMessage{Proposal: proposal}
+	return registry.sender.Send(&unregisterProducer{message: message})
+}
+
+// PingProposal pings service proposal as being alive
+func (registry *registry) PingProposal(proposal market.ServiceProposal, signer identity.Signer) error {
```
Comment (vkuznecovas) [inline core/discovery/communication/registry.go]: Do explicitly omit the signer if you're not going to use it.

File: core/discovery/docs.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Package discovery is responsible for:
+//  - handling discovery related communications
+//  - keeping track proposals in registry
+package discovery
```
Comment (vkuznecovas) [inline core/discovery/docs.go]: do we really need a new file for the 3 lines?

File: cmd/di.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Waldz) [reply cmd/di.go]: No just moved from function to struct scope

File: core/discovery/communication/registry.go (hunk 10 above)
Comment (Waldz) [reply core/discovery/communication/registry.go]: It's initial implementation, will be used in next PR

File: core/discovery/communication/registry.go (hunk 11 above)
Comment (Waldz) [reply core/discovery/communication/registry.go]: No point, will be used in next PR

File: core/discovery/docs.go (hunk 13 above)
Comment (Waldz) [reply core/discovery/docs.go]: Thats convention for package docs

File: core/discovery/communication/registry.go (hunk 12 above)
Comment (Waldz) [reply core/discovery/communication/registry.go]: No point, will be used in next PR

**Comment by Waldz** [comment]
@soffokl @tadaskay all comments addressed

File: cmd/di.go [hunk 14]
```diff
@@ -517,6 +527,25 @@ func (di *Dependencies) bootstrapIdentityComponents(options node.Options) {
 	di.IdentityRegistration = identity_registry.NewRegistrationDataProvider(di.Keystore)
 }
 
+func (di *Dependencies) bootstrapDiscoveryComponents(options node.OptionsDiscovery) error {
+	var registry discovery.ProposalRegistry
+	switch options.Type {
+	case node.DiscoveryTypeAPI:
+		registry = discovery_api.NewRegistry(di.MysteriumAPI)
+	case node.DiscoveryTypeBroker:
+		sender := discovery_communication.NewSender(nats.NewConnectionFake())
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply cmd/di.go]: renamed to mock, but it is a mock nonetheless.

File: core/discovery/interface.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/mysteriumnetwork/node/market"
+)
+
+// ProposalRegistry defines methods for proposal lifecycle - registration, keeping up to date, removal
+type ProposalRegistry interface {
+	RegisterProposal(proposal market.ServiceProposal, signer identity.Signer) error
+	PingProposal(proposal market.ServiceProposal, signer identity.Signer) error
```
[the remark this answers is not in the document]
Comment (vkuznecovas) [reply core/discovery/interface.go]: Maybe at a future date, leaving it here as it's probably good to have for the api based implementation not to be changed.

**Pull request merged**